gpsdio_segment/discrepancy.py

```diff
@@ -0,0 +1,122 @@
+import math
+
+inf = float("inf")
+
+
+class DiscrepancyCalculator(object):
+    """Base class that supplies discrepancy calculator"""
+
+    very_slow = 0.35
+
+    shape_factor = 4.0
+
+    def _update(self, key, values):
+        """Update existing instance value of `key` from value map if set
+
+        Parameters
+        ----------
+        key : str
+        values: dict
+        """
+        if key in values:
+            setattr(self, key, values[key])
+        elif not hasattr(self, key):
+            raise ValueError('instance has no default value for "{}"'.format(key))
+
+    @staticmethod
+    def compute_ts_delta_hours(ts1, ts2):
+        return (ts2 - ts1).total_seconds() / 3600
+
+    @staticmethod
+    def compute_msg_delta_hours(msg1, msg2):
+        ts1 = msg1['timestamp']
+        ts2 = msg2['timestamp']
+        return DiscrepancyCalculator.compute_ts_delta_hours(ts1, ts2)
+
+    @classmethod
+    def _compute_expected_position(cls, msg, hours):
+        epsilon = 1e-3
+        x = msg['lon']
+        y = msg['lat']
+        speed = msg['speed']
+        course = msg['course']
+        if course > 359.95:
+            assert speed <= cls.very_slow, (course, speed)
+            speed = 0
+        # Speed is in knots, so `dist` is in nautical miles (nm)
+        dist = speed * hours 
+        course = math.radians(90.0 - course)
+        deg_lat_per_nm = 1.0 / 60
+        deg_lon_per_nm = deg_lat_per_nm / (math.cos(math.radians(y)) + epsilon)
+        dx = math.cos(course) * dist * deg_lon_per_nm
+        dy = math.sin(course) * dist * deg_lat_per_nm
+        return x + dx, y + dy
+
+    def compute_discrepancy(self, msg1, msg2, hours=None):
+
+        """
+        Compute the stats required to determine if two points are continuous.  Input
+        messages must have a `lat`, `lon`, `course`, `speed` and `timestamp`, 
```

Comment:
@bitsofbits you are doing several calculations using course here. I am worried that this new version is not compatible with some VMS data. We need to make sure we all understand how the course is measured in AIS (what is taken as 0) so when we normalize VMS we convert it to the same format. 

Also, what will happen if for any reason speed or course are null?

Reply:
For AIS, north is 0 and course is measured clockwise. I will 

When speed or course is null, those messages are kicked out as `BAD_MESSAGE` by
`_message_type`  If VMS doesn't have those, we'll need to do something different here -- for instance we could pass a flag in to do the computation without course and speed.